### PR TITLE
Value binding with, addendum

### PR DESCRIPTION
**Pull request by peterbourgon**
Cosmetic, documentation, and organizational changes. No change in functionality.

### CODE REVIEW

**Comment by peterbourgon** [comment]
@ChrisHines, can you take a look? @tsenart, I believe this, particularly 2348d79, addresses your global state concerns?

**Comment by tsenart** [comment]
LGMT though I think that a `ValuerFunc` would still be useful for users of the package.

**Comment by peterbourgon** [comment]
I tried using a ValuerFunc, but I felt like it just added ceremony and SLOC, without concomitant value...

File: log/value_test.go [hunk 1]
```diff
@@ -10,26 +10,26 @@ import (
...
 	var output []interface{}
+
 	logger := log.Logger(log.LoggerFunc(func(keyvals ...interface{}) error {
 		output = keyvals
 		return nil
 	}))
-	logger = log.With(logger, "ts", log.Timestamp, "caller", log.Caller)
+	logger = log.With(logger, "ts", log.DefaultTimestamp, "caller", log.DefaultCaller)
 
-	time.Sleep(50 * time.Millisecond)
```
Comment (ChrisHines) [inline log/value_test.go:19]: This Sleep was present to make sure that the time was the time of the Log call, not the time of the With call. Note that on Windows time is generally reported at maximum granularity of 1ms, and sometimes no better than 15ms. Also on Windows time.Sleep may not have any effect below 15ms.

**Comment by ChrisHines** [comment]
LGTM, after addressing my inline comment above about the time.Sleep in the test code.

**Comment by tsenart** [comment]
@peterbourgon: I think it's a good trade-off for generality. https://gist.github.com/tsenart/80fdfa09206d5d46f1ca

**Comment by ChrisHines** [comment]
@tsenart  I tend to agree with @peterbourgon. It seems like something we could add later if it feels missing after we've been using the package for a while.

**Comment by tsenart** [comment]
Alright, sure. I don't wanna block progress for this. But I'll leave this discussion with this idea: https://gist.github.com/tsenart/377563923d300c5a5f64

**Comment by ChrisHines** [comment]
@tsenart, how would `BindValues` work without the `Valuer` interface?

**Comment by tsenart** [comment]
@ChrisHines: We're still doing a type assertion on the `Valuer` type, being an interface or a function doesn't matter.

**Comment by tsenart** [comment]
http://golang.org/ref/spec#Type_assertions

**Comment by ChrisHines** [comment]
@tsenart, you are right. I was thinking backwards about which side of the type assertion must be an interface. We are asserting from `interface{}` so all is good.

As I said before, I don't have a strong opinion on this topic, just a preference for waiting when I'm not sure. On the other hand I don't think we are ready to lock down the API soon, so I am happy with either choice.

**Comment by ChrisHines** [comment]
@tsenart On further thought, is the `Value` type still needed in your latest gist? Can we get away with the following?

``` go
type Valuer func() interface{}
```

There is no chance for an arbitrary function to be mistaken for a `Valuer` since `Valuer` is a named function type instead of an interface. A plain `func () interface{}` will not accidentally match the type assertion to `Valuer` in `BindValues`, so the reason for creating type `Value` disappears.

This factor swings my opinion in favor of your idea.

**Comment by tsenart** [comment]
Good observation!

**Comment by ChrisHines** [comment]
I think this is the way to go (no pun intended). When I was conceiving this API I knew the core of the idea was to store a value generating function in the log context, but I didn't follow through on that idea far enough. I'm glad you stuck to your guns.

Unless @peterbourgon objects, I'll merge his changes and your tweak into my original PR later tonight (my time) and we can take the review back to the main repo.

File: log/value_test.go (hunk 1 above)
Comment (peterbourgon) [inline log/value_test.go:19]: Can we do 20ms?

**Comment by tsenart** [comment]
Open source works :-)

File: log/value_test.go (hunk 1 above)
Comment (ChrisHines) [inline log/value_test.go:19]: I will figure out a smaller value that still works after I merge the changes from this discussion.

**Comment by peterbourgon** [comment]
:dango:

File: log/value_test.go (hunk 1 above)
Comment (tsenart) [inline log/value_test.go:19]: Well, now you can mock the time function so no need to actually sleep!

On Friday, April 24, 2015, Chris Hines notifications@github.com wrote:

> In log/value_test.go
> https://github.com/ChrisHines/gokit/pull/1#discussion_r29079367:
> 
> > -   time.Sleep(50 \* time.Millisecond)
> 
> I will figure out a smaller value that still works after I merge the
> changes from this discussion.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ChrisHines/gokit/pull/1/files#r29079367.

## 

Sent from Gmail Mobile

Comment (ChrisHines) [inline log/value_test.go:19]: True. And in reality the test just needs to validate that the function gets called from the right place, so the time value doesn't actually matter. The test can inject a function that increments a counter and check that the increment happens only on Log calls.

Comment (tsenart) [inline log/value_test.go:19]:
``` go
func clock(start time.Time, step time.Duration) func() time.Time {
    var steps int64
    return func() time.Time {
        return start.Add(time.Duration(atomic.AddInt64(&steps, 1) * int64(step)))
    }
}
```

**Pull request merged**